packages/opencensus-core/src/trace/model/span.ts

```diff
@@ -13,62 +13,354 @@
+  attributes: types.Attributes = {};
+  /** A text annotation with a set of attributes. */
+  annotations: types.Annotation[] = [];
+  /** An event describing a message sent/received between Spans */
+  messageEvents: types.MessageEvent[] = [];
+  /** Pointers from the current span to another span */
+  links: types.Link[] = [];
+  /** If the parent span is in another process. */
+  remoteParent: boolean;
+  /** This span's root span.  If it's a root span, it will point to this */
+  root: RootSpan;
+  /** This span's parent. If it's a root span, must be empty */
+  parentSpan: Span = null;
+  /** The resource name of the span */
+  name: string = null;
+  /** Kind of span. */
+  kind: types.SpanKind = types.SpanKind.UNSPECIFIED;
+  /** A final status for this span */
+  status: types.Status = STATUS_OK;
+  /** Trace Parameters */
+  activeTraceParams: configTypes.TraceParams;
 
-  /**
-   * Constructs a new SpanImpl instance.
-   * @param root
-   */
-  constructor(root: types.RootSpan) {
-    super();
-    this.root = root;
-    this.logger = this.root.logger || logger.logger();
-    this.parentSpanId = root.id;
-    this.activeTraceParams = this.root.activeTraceParams;
-  }
+  /** The number of dropped attributes. */
+  droppedAttributesCount = 0;
+  /** The number of dropped links. */
+  droppedLinksCount = 0;
+  /** The number of dropped annotations. */
+  droppedAnnotationsCount = 0;
+  /** The number of dropped message events. */
+  droppedMessageEventsCount = 0;
 
-  /** Gets trace id of span. */
+  /** Constructs a new Span instance. */
+  constructor(parent: Span) {
+    this.className = this.constructor.name;
+    this.id = randomSpanId();
+    this.spansLocal = [];
+    if (parent) {
+      this.root = parent.root;
+      this.parentSpan = parent;
+    }
+    this.logger = (this.root && this.root.logger) || logger.logger();
+  }
+ 
+  /** Gets the trace ID. */
   get traceId(): string {
     return this.root.traceId;
   }
 
-  get traceState(): string {
+  /** Gets the trace state */
+  get traceState(): types.TraceState {
     return this.root.traceState;
   }
 
-  /** Starts the span instance. */
-  start() {
-    super.start();
-    this.logger.debug('starting span  %o', {
+  /** Gets the ID of the parent span. */
+  get parentSpanId(): string {
+    return this.parentSpan.id;
+  }
+
+  /** Indicates if span was started. */
+  get started(): boolean {
+    return this.startedLocal;
+  }
+
+  /** Indicates if span was ended. */
+  get ended(): boolean {
+    return this.endedLocal;
+  }
+
+  /**
+   * Gives a timestamp that indicates the span's start time in RFC3339 UTC
+   * "Zulu" format.
+   */
+  get startTime(): Date {
+    if (!this.clock) {
+      this.logger.debug('calling startTime() on null clock');
+      return null;
+    }
+
+    return this.clock.startTime;
+  }
+
+  /** Recursively gets the descendent spans. */
+  get spans(): types.Span[] {
+    return this.spansLocal.reduce((acc: types.Span[], cur) => {
```

Comment:
I think this approach is OK and keeps compatibility with the existing exporters, but part of me feels like this logic should live in the exporters or exporter buffer so that this function could just give the immediate children.

Reply:
Yes, I did it for compatibility with the existing exporters.  I could make this one just return the immediate children and add `allDescendents()` to do the recursive version.  Then we could update the exporters with a one line change.

Reply:
Ah, I like that approach!